rfd/0015-2fa-management.md

```diff
@@ -0,0 +1,206 @@
+device enrolled. The [bootstrap](#bootstrap) section covers registration of the
+first device.
+
+#### CLI
+
+Login:
+
+```sh
+$ tsh login --login=non-sso-user
+Enter password for Teleport user non-sso-user: ...
+Tap your security key... <tap>
+
+$ tsh login --login=non-sso-user --2fa=otp
+Enter password for Teleport user non-sso-user: ...
+Enter OTP code: ...
+
+$ tsh login --login=sso-user --auth=oidc
+# SSO page opens
+# no 2FA prompt from Teleport
+```
+
+2FA management:
+
+```sh
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+android OTP        OTP    Tue 15 Dec 2020 01:29:42 PM PST
+yubikey            U2F    Wed 16 Dec 2020 02:00:13 PM PST
+
+$ tsh 2fa add
+Adding a new 2FA device.
+Choose device type (1 - OTP, 2 - U2F): 2
+Enter device name: solokey
+Tap any *registered* security key... <tap>
+Tap your *new* security key... <tap>
+2FA device "solokey" added.
+
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+android OTP        OTP    Tue 15 Dec 2020 01:29:42 PM PST
+yubikey            U2F    Wed 16 Dec 2020 02:00:13 PM PST
+solokey            U2F    Wed 16 Dec 2020 02:05:46 PM PST
+
+$ tsh 2fa rm yubikey
+Tap any *registered* security key... <tap>
+2FA device "yubikey" removed.
+
+$ tsh 2fa rm "android OTP"
+Tap any *registered* security key... <tap>
+2FA device "android OTP" removed.
+
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+solokey            U2F    Wed 16 Dec 2020 02:06:46 PM PST
+
+# If 2FA is optional:
+$ tsh 2fa rm solokey
+You are about to remove the only remaining 2FA device.
+This will disable 2FA during login.
+Are you sure? (y/N): N
+
+# If 2FA is required:
+$ tsh 2fa rm solokey
+Can't remove the only remaining 2FA device.
+Please add a replacement 2FA device first using "tsh 2fa add".
+```
+
+#### web UI
+
+Web UI management of 2FA devices should be logically similar to the CLI:
+- a page to see all enrolled devices
+- a button to enroll a new device
+- buttons to remove any enrolled device (except for the last one)
+
+Web UI details, wireframes and implementation will be added later, when we have
+the capacity to do it. Initially, 2FA management is CLI-only.
+
+#### bootstrap
+
+Initially, a user account doesn't have a 2FA device. Depending on [cluster
+configuration](#configuration), use of 2FA devices might be required.
+
+If 2FA is required, a user is required to enroll a device during account
+creation (for local users) or first login (for SSO users).
+
+If 2FA is optional, a user can create an account and login without 2FA. They
+can then add 2FA devices as described above. If an existing user has at least
+one 2FA device registered, it's required during login.
+
+### configuration
+
+The current 2FA configuration in Teleport only applies to local users and is
```

Comment:
The biggest challenge is to let SSO users enroll their keys. These users are ephemeral, so if we connect any data under the username, it will get deleted. I haven't figured out how to combine those two flows, but this feature will be used by Enterprise users.

Reply:
I remember asking this in the past, but don't remember the answer: can we simply stop expiring SSO users from our backend?

Reply:
If we do this by default, this could lead to confusion, because administrators would need to start clean those entries up periodically. 

If someone enrolls a U2F key, we can remove TTL from this SSO user and make sure it won't be set. So we will treat those users differently.

Reply:
Are Teleport-stored users exposed anywhere admin-facing? AFAIK it's just about backend storage consumption, right?

A long TTL, like 1-2 years, could be a good balance between accumulating backend garbage and not deleting users after prolonged inactivity. Worst case: someone doesn't touch teleport for a year and after a fresh SSO login they need to re-add U2F keys.

Reply:
they are visible in the same UI alongside the local users.

Reply:
we also store MFA data separately from other user data
we could persist MFA data (with very long TTL) and keep other user data ephemeral as it is today